lib/date_component.rb

```diff
@@ -1,44 +1,85 @@
+        $logger.debug 'Parsing date strings from values', { values: @date_values }
+        # Split date strings into start/end values and then pivot them into properly arranged arrays
+        # e.g. [[1, 2], [3, 4], [5, 6]] to [[1, 3, 5], [2, 4, 6]]
+        date_components = @date_values.map { |v| _extract_date_components v }.transpose
+
+        begin
+            start_str = _transform_date_components_to_str date_components[0]
+            end_str = _transform_date_components_to_str date_components[1]
+            $logger.info "Setting date values: start / #{start_str}, end / #{end_str}"
 
-        if start_str == end_str
-            { start: start_str, end: nil }
-        else
-            { start: start_str, end: end_str }
+            @date_strs[:start] = start_str
+            @date_strs[:end] = end_str && end_str != start_str ? end_str : nil
+        rescue Date::Error, TypeError
+            $logger.error 'Unable to parse date values for this check-in box'
+            $logger.debug date_components
         end
     end
 
     private
 
-    def _format_str(pos)
-        year = instance_variable_get("@#{pos}_year")
-        month = instance_variable_get("@#{pos}_month") || '-'
-        day = instance_variable_get("@#{pos}_day")
+    def _extract_date_components(component)
+        component_arr = component.split('-')
+        component_arr[1] ? component_arr : [component_arr[0], component_arr[0]]
+    end
+
+    # rubocop:disable Metrics/CyclomaticComplexity
+    def _transform_date_components_to_str(part)
+        $logger.debug 'Creating date string from', { values: part }
+        date_str = nil
+
+        # Create a simplified array with the shape of the date values
+        # This gives us a good idea of what the date format is
+        null_positions = part.map { |x| x.nil? ? 0 : 1 }
+
+        # Convert to array of ints with no nil values since DateTime doesn't like them
+        date_array = part.map { |x| x.nil? ? x : x.to_i }
+
+        case null_positions
+        # Single value should be a year and returned as-is
+        when [1, 0, 0]
+            date_str = date_array[0].to_s
+        # Two consecutive values should be Month-Year or Season-Year
+        when [1, 1, 0]
+            begin
+                date_str = DateTime.new(*date_array.compact).strftime('%b. %Y')
+            rescue Date::Error => e
+                $logger.error 'Unable to parse date', { value: date_array, reason: e }
+                # DateTime doesn't know about seasons, so fake it here
+                date_str = "#{@@season_codes[date_array[1]]} #{date_array[0]}"
+            end
+        # Three values will 99.99% of the time be a true date
+        when [1, 1, 1]
+            date_str = DateTime.new(*date_array).strftime('%b. %-d, %Y')
+        # A missing value in the month/season position might indicate a day-of-year value
+        when [1, 0, 1]
+            date_str = "Day #{date_array[2]} of #{date_array[0]}"
+        end
```

Comment:
Are we sure we don't have to handle any other possibilities in [0,1]^3 ?

Reply:
Pretty sure, not 100%, but I've looked at the source data, and there are about 12 rows out of 800k that don't conform to these arrangements and all of those are nonsense in the context of being understood of as a date.

Essentially everything else in that space doesn't make sense as a date for various reasons, mainly that `[0, *, *]` doesn't work because we always need a year for a valid date. It's _possible_ its present in another position, but I haven't seen it and we should treat it as a mis-cataloged record rather than something we have to deal with if it does come up.

Reply:
In that case should we raise an error if it doesn't match one of the expected formats?

Reply:
I've added an error to be raised there, but am catching it in the `create_strs` method since if we can't parse the date I still want to process the rest of the check-in card. But this way if people notice there is data missing we'll have some nicely logged errors explaining what went wrong.